### PR TITLE
`GetLiveness` API endpoint

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

Implements the https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Validator/getLiveness API endpoint.

**Other notes for review**

I got the idea of using participation bits from @nisdas. We already use them for doppelganger check in Prysm APIs: https://github.com/prysmaticlabs/prysm/blob/8ade8afb73c90f0847c51e47517b7825f56c9b1c/beacon-chain/rpc/prysm/v1alpha1/validator/status.go#L101

### CODE REVIEW

File: beacon-chain/rpc/eth/validator/validator_test.go [hunk 1]
```diff
@@ -3866,3 +3866,69 @@ func TestServer_SubmitValidatorRegistrations(t *testing.T) {
...
+
+	t.Run("current epoch", func(t *testing.T) {
+		resp, err := server.GetLiveness(ctx, &ethpbv2.GetLivenessRequest{
+			Epoch:   0,
+			Indices: []types.ValidatorIndex{0, 1},
+		})
+		require.NoError(t, err)
+		data0 := resp.Data[0]
+		data1 := resp.Data[1]
+		assert.Equal(t, true, (data0.Index == 0 && !data0.IsLive) || (data0.Index == 1 && !data0.IsLive))
```
Comment (rkapka) [inline beacon-chain/rpc/eth/validator/validator_test.go:3902]: I don't assume any ordering

File: beacon-chain/rpc/apimiddleware/structs.go [hunk 2]
```diff
@@ -213,7 +213,7 @@ type SpecResponseJson struct {
 	Data interface{} `json:"data"`
 }
 
-type DutiesRequestJson struct {
+type ValidatorIndicesJson struct {
```
Comment (rkapka) [inline beacon-chain/rpc/apimiddleware/structs.go:216]: I changed the name to be more reusable.

File: beacon-chain/rpc/eth/validator/validator.go [hunk 3]
```diff
@@ -37,6 +37,7 @@ import (
 )
 
 var errInvalidValIndex = errors.New("invalid validator index")
+var errParticipation = status.Errorf(codes.Internal, "Could not obtain epoch participation")
```
Comment (prestonvanloon) [inline beacon-chain/rpc/eth/validator/validator.go:40]: There is no format here, prefer `status.Error`
```suggestion
var errParticipation = status.Error(codes.Internal, "Could not obtain epoch participation")
```

File: beacon-chain/rpc/eth/validator/validator.go [hunk 4]
```diff
@@ -966,6 +967,70 @@ func (vs *Server) SubmitContributionAndProofs(ctx context.Context, req *ethpbv2.
...
+	// If it is, then we won't be able to fetch the state at the end of the epoch.
+	// In that case we get participation info from the head state.
+	// We can also use the head state to get participation info for the previous epoch.
+	headSt, err := vs.HeadFetcher.HeadState(ctx)
+	if err != nil {
+		return nil, status.Error(codes.Internal, "Could not get head state")
+	}
+	currEpoch := slots.ToEpoch(headSt.Slot())
+	if req.Epoch > currEpoch {
+		return nil, status.Errorf(codes.InvalidArgument, "Requested epoch cannot be in the future")
```
Comment (prestonvanloon) [inline beacon-chain/rpc/eth/validator/validator.go:993]:
```suggestion
		return nil, status.Error(codes.InvalidArgument, "Requested epoch cannot be in the future")
```

File: beacon-chain/rpc/eth/validator/server.go [hunk 5]
```diff
@@ -26,4 +27,5 @@ type Server struct {
 	SyncCommitteePool      synccommittee.Pool
 	V1Alpha1Server         *v1alpha1validator.Server
 	ProposerSlotIndexCache *cache.ProposerPayloadIDsCache
+	ReplayerBuilder        stategen.ReplayerBuilder
```
Comment (prestonvanloon) [inline beacon-chain/rpc/eth/validator/server.go:30]: What is this used for? I don't see it used in this PR.

**Review by prestonvanloon — CHANGES_REQUESTED** [review]

File: beacon-chain/rpc/eth/validator/validator.go [hunk 6]
```diff
@@ -966,6 +967,70 @@ func (vs *Server) SubmitContributionAndProofs(ctx context.Context, req *ethpbv2.
...
+		}
+	}
+
+	resp := &ethpbv2.GetLivenessResponse{
+		Data: make([]*ethpbv2.GetLivenessResponse_Liveness, len(req.Index)),
+	}
+	for i, vi := range req.Index {
+		resp.Data[i] = &ethpbv2.GetLivenessResponse_Liveness{
+			Index:  vi,
+			IsLive: participation[vi] != 0,
```
Comment (prestonvanloon) [inline beacon-chain/rpc/eth/validator/validator.go:1027]: Could this panic if vi is outside of the range of `len(participation)` ?

Comment (prestonvanloon) [reply beacon-chain/rpc/eth/validator/validator.go:1027]: If so, add a regression test please.

Comment (rkapka) [reply beacon-chain/rpc/eth/validator/validator.go:1037]: Yes, if the validator index is invalid. I added a defensive check.

File: beacon-chain/rpc/eth/validator/validator.go [hunk 7]
```diff
@@ -1022,6 +1025,13 @@ func (vs *Server) GetLiveness(ctx context.Context, req *ethpbv2.GetLivenessReque
 		Data: make([]*ethpbv2.GetLivenessResponse_Liveness, len(req.Index)),
 	}
 	for i, vi := range req.Index {
+		_, err = st.ValidatorAtIndex(vi)
```
Comment (prestonvanloon) [inline beacon-chain/rpc/eth/validator/validator.go:1028]: This is a bit of an expensive operation. It makes a copy of the validator and if you are iterating over a large request then you might have a noticeable burden on the server.

I think you could simply check that `vi` is less than `len(participation)` and return the out of range error if it fails that check

**Review by prestonvanloon — APPROVED** [review]
Nice, I think this would work.

I didn't test it locally, but the logic makes sense and the code is good. Nice work

**Pull request merged**